File: js/packages/teams-ai/src/TeamsAttachmentDownloader.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import axios, { AxiosInstance } from 'axios';
import { Attachment, TurnContext } from 'botbuilder';
import { AppCredentials, AuthenticationConstants, GovernmentConstants } from 'botframework-connector';

import { InputFile, InputFileDownloader } from './InputFileDownloader';
import { TeamsAdapter } from './TeamsAdapter';
import { TurnState } from './TurnState';

/**
 * Options for the `TeamsAttachmentDownloader` class.
 */
export interface TeamsAttachmentDownloaderOptions {
    /**
     * The Microsoft App ID of the bot.
     */
    botAppId: string;

    /**
     * ServiceClientCredentialsFactory
     */
    adapter: TeamsAdapter;
}

export interface AuthenticatorResult {
    /**
     * The value of the access token resulting from an authentication process.
     */
    accessToken: string;
    /**
     *  The date and time of expiration relative to Coordinated Universal Time (UTC).
     */
    expiresOn: Date;
}
/**
 * Downloads attachments from Teams using the bots access token.
 */
export class TeamsAttachmentDownloader<TState extends TurnState = TurnState> implements InputFileDownloader<TState> {
    private readonly _options: TeamsAttachmentDownloaderOptions;
    private _httpClient: AxiosInstance;

    /**
     * Creates a new instance of the `TeamsAttachmentDownloader` class.
     * @param {TeamsAttachmentDownloader} options - Options for the `TeamsAttachmentDownloader` class.
     */
    public constructor(options: TeamsAttachmentDownloaderOptions) {
        this._options = options;
        this._httpClient = axios.create();
    }

    /**
     * Download any files relative to the current user's input.
     * @template TState - Type of the state object passed to the `TurnContext.turnState` method.
     * @param {TurnContext} context Context for the current turn of conversation.
     * @param {TState} state Application state for the current turn of conversation.
     * @returns {Promise<InputFile[]>} Promise that resolves to an array of downloaded input files.
     */
    public async downloadFiles(context: TurnContext, state: TState): Promise<InputFile[]> {
        // Filter out HTML attachments
        const attachments = context.activity.attachments?.filter((a) => !a.contentType.startsWith('text/html'));
        if (!attachments || attachments.length === 0) {
            return Promise.resolve([]);
        }

        let accessToken = '';

        // If authentication is enabled, get access token
        if ((await this._options.adapter.credentialsFactory?.isAuthenticationDisabled()) !== true) {
            // Download all attachments
            accessToken = await this.getAccessToken();
        }
        const files: InputFile[] = [];
        for (const attachment of attachments) {
            const file = await this.downloadFile(attachment, accessToken);
            if (file) {
                files.push(file);
            }
        }

        return files;
    }

    /**
     * @private
     * @param {Attachment} attachment - Attachment to download.
     * @param {string} accessToken - Access token to use for downloading.
     * @returns {Promise<InputFile>} - Promise that resolves to the downloaded input file.
     */
    private async downloadFile(attachment: Attachment, accessToken: string): Promise<InputFile | undefined> {
        if (attachment.content) {
            return {
                content: Buffer.from(attachment.content),
                contentType: attachment.contentType,
                contentUrl: attachment.contentUrl
            };
        } else if (
            (attachment.contentUrl && attachment.contentUrl.startsWith('https://')) ||
            (attachment.contentUrl && attachment.contentUrl.startsWith('http://localhost'))
        ) {
            let headers;
            if (accessToken.length > 0) {
                // Build request for downloading file if access token is available
                headers = {
                    Authorization: `Bearer ${accessToken}`
                };
            }
            const response = await this._httpClient.get(attachment.contentUrl, {
                headers,
                responseType: 'arraybuffer'
            });

            // Convert to a buffer
            const content = Buffer.from(response.data, 'binary');

            // Fixup content type
            let contentType = attachment.contentType;
            if (contentType === 'image/*') {
                contentType = 'image/png';
            }

            // Return file
            return {
                content,
                contentType,
                contentUrl: attachment.contentUrl
            };
        }
    }

    /**
     * @private
     * @returns {Promise<string>} - Promise that resolves to the access token.
     */
    private async getAccessToken(): Promise<string> {
        // Normalize the ToChannelFromBotLoginUrl (and use a default value when it is undefined).
        const toChannelFromBotLoginUrl = (
            this._options.adapter.botFrameworkAuthConfig?.ToChannelFromBotLoginUrl ||
            AuthenticationConstants.ToChannelFromBotLoginUrlPrefix + AuthenticationConstants.DefaultChannelAuthTenant
        ).toLowerCase();

        let audience = this._options.adapter.botFrameworkAuthConfig?.ToChannelFromBotOAuthScope;
        const loginEndpoint = toChannelFromBotLoginUrl;

        // If there is no loginEndpoint set on the provided ConfigurationBotFrameworkAuthenticationOptions, or it starts with 'https://login.microsoftonline.com/', the bot is operating in Public Azure.
        // So we use the Public Azure audience or the specified audience.
        if (loginEndpoint.startsWith(AuthenticationConstants.ToChannelFromBotLoginUrlPrefix)) {
            audience = audience ?? AuthenticationConstants.ToChannelFromBotOAuthScope;
        } else if (toChannelFromBotLoginUrl === GovernmentConstants.ToChannelFromBotLoginUrl) {
            // Or if the bot is operating in US Government Azure, use that audience.
            audience = audience ?? GovernmentConstants.ToChannelFromBotOAuthScope;
        }

        const appCreds = (await this._options.adapter.credentialsFactory.createCredentials(
            this._options.botAppId,
            audience,
            loginEndpoint,
            true
        )) as AppCredentials;

        return appCreds.getToken();
    }
}

File: python/packages/ai/teams/teams_attachment_downloader/teams_attachment_downloader.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

from typing import List, Optional

import aiohttp
from botbuilder.core import TurnContext
from botbuilder.schema import Attachment
from botframework.connector.auth import AuthenticationConstants, GovernmentConstants

from ..input_file import InputFile, InputFileDownloader
from .teams_attachment_downloader_options import TeamsAttachmentDownloaderOptions


class TeamsAttachmentDownloader(InputFileDownloader):
    """
    Downloads attachments from Teams using the bot's access token.
    """

    _options: TeamsAttachmentDownloaderOptions

    def __init__(self, options: TeamsAttachmentDownloaderOptions):
        """
        Creates a new instance of the 'TeamsAttachmentDownloader' class

        Args:
            options (TeamsAttachmentDownloaderOptions): The options for configuring the class.
        """
        self._options = options

    async def download_files(self, context: TurnContext) -> List[InputFile]:
        """
        Download any files relative to the current user's input.

        Args:
            context (TurnContext): Context for the current turn of conversation.

        Returns:
            List[InputFile]: The list of input files
        """

        # Filter out HTML attachments
        valid_attachments = []
        attachments = context.activity.attachments

        if attachments is None or len(attachments) == 0:
            return []

        for attachment in attachments:
            if not attachment.content_type.startswith("text/html"):
                valid_attachments.append(attachment)

        if len(valid_attachments) == 0:
            return []

        access_token = ""

        # If authentication is enabled, get access token
        if await self._options.adapter.credentials_factory.is_authentication_disabled() is False:
            access_token = await self._get_access_token()

        files: List[InputFile] = []
        for attachment in valid_attachments:
            file = await self._download_file(attachment, access_token)
            if file is not None:
                files.append(file)

        return files

    async def _download_file(
        self, attachment: Attachment, access_token: str
    ) -> Optional[InputFile]:
        valid_http = attachment.content_url is not None and attachment.content_url.startswith(
            "https://"
        )
        valid_local_host = attachment.content_url is not None and attachment.content_url.startswith(
            "http://localhost"
        )

        if valid_http or valid_local_host:
            headers = {}

            if len(access_token) > 0:
                # Build request for downloading file if access token is available
                headers.update({"Authorization": f"Bearer {access_token}"})

            download_url = attachment.content_url
            if attachment.content and isinstance(attachment.content, dict):
                download_url = attachment.content.get("downloadUrl", attachment.content_url)
            async with aiohttp.ClientSession() as session:
                async with session.get(download_url, headers=headers) as response:
                    content = await response.read()

                    content_type = attachment.content_type

                    if content_type == "image/*":
                        content_type = "image/png"

                    return InputFile(content, content_type, attachment.content_url)
        else:
            content = bytes(attachment.content) if attachment.content else bytes()
            return InputFile(content, attachment.content_type, attachment.content_url)

    async def _get_access_token(self):
        # Normalize the to_channel_from_bot_login_url_prefix (and use a default when undefined).
        # If non-public (specific tenant) login URL is to be used, make sure the full url
        # including tenant ID is provided to TeamsAdapter on setup.
        to_channel_from_bot_login_url_default = (
            AuthenticationConstants.TO_CHANNEL_FROM_BOT_LOGIN_URL_PREFIX
            + AuthenticationConstants.DEFAULT_CHANNEL_AUTH_TENANT
        )

        to_channel_from_bot_login_url = getattr(
            self._options.adapter.configuration, "TO_CHANNEL_FROM_BOT_LOGIN_URL", ""
        )

        if not to_channel_from_bot_login_url:
            to_channel_from_bot_login_url = to_channel_from_bot_login_url_default

        audience = getattr(
            self._options.adapter.configuration, "TO_CHANNEL_FROM_BOT_OAUTH_SCOPE", ""
        )

        # If there is no to_channel_from_bot_login_url set on the provided
        # ConfigurationBotFrameworkAuthenticationOptions, or it starts with
        # 'https://login.microsoftonline.com/', the bot is operating in
        # Public Azure. So we use the Public Azure audience
        # or the specified audience.
        if to_channel_from_bot_login_url.startswith(
            AuthenticationConstants.TO_CHANNEL_FROM_BOT_LOGIN_URL_PREFIX
        ):
            if not audience:
                audience = AuthenticationConstants.TO_CHANNEL_FROM_BOT_OAUTH_SCOPE
        elif to_channel_from_bot_login_url.startswith(
            GovernmentConstants.TO_CHANNEL_FROM_BOT_LOGIN_URL_PREFIX
        ):
            # Or if the bot is operating in US Government Azure, use that audience.
            if not audience:
                audience = GovernmentConstants.TO_CHANNEL_FROM_BOT_OAUTH_SCOPE

        app_creds = await self._options.adapter.credentials_factory.create_credentials(
            self._options.bot_app_id, audience, to_channel_from_bot_login_url, True
        )
        return app_creds.get_access_token()
